qos/solana/context.go

```diff
@@ -0,0 +1,187 @@
+		Data: string(reqBz),
+		// Method is alway POST for Solana.
+		Method: http.MethodPost,
+
+		// Path field is not used for Solana.
+
+		// TODO_IMPROVE: adjust the timeout based on the request method:
+		// An endpoint may need more time to process certain requests,
+		// as indicated by the request's method and/or parameters.
+		TimeoutMillisec: defaultServiceRequestTimeoutMillisec,
+	}
+}
+
+// UpdateWithResponse is NOT safe for concurrent use
+func (rc *requestContext) UpdateWithResponse(endpointAddr relayer.EndpointAddr, responseBz []byte) {
+	// TODO_IMPROVE: check whether the request was valid, and return an error if it was not.
+	// This would be an extra safety measure, as the caller should have checked the returned value
+	// indicating the validity of the request when calling on QoS instance's ParseHTTPRequest
+
+	response, err := unmarshalResponse(rc.JSONRPCReq, responseBz, rc.Logger)
+
+	// TODO_UPNEXT(@adshmh): Drop the unmarshalling error: the returned response interface should provide methods to allow the caller to:
+	// 1. Check if the response from the endpoint was valid or malformed. This is needed to support retrying with a different endpoint if
+	// the originally selected one fails to return a valid response to the user's request.
+	// 2. Return a generic but valid JSONRPC response to the user.
+	rc.endpointResponses = append(rc.endpointResponses,
+		endpointResponse{
+			EndpointAddr: endpointAddr,
+			response:     response,
+			unmarshalErr: err,
+		},
+	)
+}
+
+// GetHTTPResponse builds the HTTP response that should be returned for
+// a Solana blockchain service request.
+func (rc requestContext) GetHTTPResponse() gateway.HTTPResponse {
+	var response response
+
+	if len(rc.endpointResponses) >= 1 {
+		// return the last endpoint response reported to the context.
+		response = rc.endpointResponses[len(rc.endpointResponses)-1]
+	} else {
+		// By default, return a generic HTTP response if no endpoint responses
+		// have been reported to the request context.
+		// intentionally ignoring the error here, since unmarshallResponse
+		// is being called with an empty endpoint response payload.
+		response, _ = unmarshalResponse(rc.JSONRPCReq, []byte(""), rc.Logger)
+	}
+
+	return httpResponse{
+		responsePayload: response.GetResponsePayload(),
+	}
+}
+
+func (rc requestContext) GetObservationSet() message.ObservationSet {
+	// No updates needed if the request was invalid
+	if !rc.isValid {
+		return observationSet{}
+	}
+
+	observations := make(map[relayer.EndpointAddr][]observation)
+	for _, response := range rc.endpointResponses {
+		obs, ok := response.GetObservation()
+		if !ok {
+			continue
+		}
+
+		addr := response.EndpointAddr
+		observations[addr] = append(observations[addr], obs)
+	}
+
+	return observationSet{
+		EndpointStore: rc.EndpointStore,
+		ServiceState:  rc.ServiceState,
+		Observations:  observations,
+	}
+}
+
+func (rc *requestContext) GetEndpointSelector() relayer.EndpointSelector {
```

Comment:
Can you add a comment describing that the `requestContext` satisfies the `relayer.EndpointSelector` interface and why that is.